### ISSUE TITLE
Experiment results of Cacheblend

### ISSUE THREAD

**Issue by lhr-30** [UCSB-NLP-Chang/KVLink#1]
Hi, I have seen that you provide the results of Cacheblend in your experiment part. May I ask how you reproduce their results? 
It seems that CacheBlend doesn't support rope scaling. I tried to delete the rope scaling para in config.json using llama-3.2-1B but get errors.

**Comment by KimperYang** [comment]
Hi, thank you for your interest in our work. We encountered the same issue of not supporting rope scaling with LMCache.

Besides the LMCache repo, there's another CacheBlend repo released by the authors earlier. We followed their modification on the vLLM in that repo and adpated the same changes to the huggingface Llama.

When reproducing CacheBlend, we used the position_ids arguments to directly give all the reused cache with the correct position during the forward pass to avoid the rope scaling.

Specifically, we implement the following function based on the Huggingface Llama and CacheBlend implementation in their original repo:

```python
def do_blend(model, old_kv, golden_kv, recompute_ratio, first_layer_states, position_ids, config):

    blended_kv = []
    blended_kv.append(golden_kv[0])
    blended_kv.append(golden_kv[1])
    sequence_length = old_kv[0][0].size(2)
    topk_num = int(sequence_length * recompute_ratio)

    first_layer_old_value = old_kv[1][1]
    first_layer_golden_value = golden_kv[1][1]

    temp_diff = torch.sum((first_layer_golden_value - first_layer_old_value)**2, dim=[0,1,3]) #remain the sequence length dimension
    top_indices = torch.topk(temp_diff, k=topk_num).indices
    top_indices, _ = torch.sort(top_indices)
    # print(top_indices)

    # if "head_dim" in config.keys():
    #     head_dim = config.head_dim
    # else:
    head_dim = config.hidden_size // config.num_attention_heads
    for layer_idx in range(2, len(old_kv)):
        if layer_idx == 2:
            hidden_states = first_layer_states
            hidden_states = hidden_states[:, top_indices, :]
            position_ids = position_ids[:, top_indices]


        residual = hidden_states

        transformer_block = model.model.layers[layer_idx]

        hidden_states = transformer_block.input_layernorm(hidden_states)

        bsz, q_len, _ = hidden_states.size()

        query_states = transformer_block.self_attn.q_proj(hidden_states)
        key_states = transformer_block.self_attn.k_proj(hidden_states)
        value_states = transformer_block.self_attn.v_proj(hidden_states)

        # use -1 to infer num_heads and num_key_value_heads as they may vary if tensor parallel is used
        query_states = query_states.view(bsz, q_len, -1, head_dim).transpose(1, 2)
        key_states = key_states.view(bsz, q_len, -1, head_dim).transpose(1, 2)
        value_states = value_states.view(bsz, q_len, -1, head_dim).transpose(1, 2)

        cos, sin = transformer_block.self_attn.rotary_emb(value_states, position_ids)
        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)

        updated_old_k = old_kv[layer_idx][0].clone()
        updated_old_v = old_kv[layer_idx][1].clone()

        updated_old_k[:, :, top_indices, :] = key_states
        updated_old_v[:, :, top_indices, :] = value_states

        blended_kv.append((updated_old_k, updated_old_v))

        key_states = repeat_kv(updated_old_k, transformer_block.self_attn.num_key_value_groups)
        value_states = repeat_kv(updated_old_v, transformer_block.self_attn.num_key_value_groups)

        attn_output = torch.nn.functional.scaled_dot_product_attention(
            query_states,
            key_states,
            value_states,
            attn_mask=None,
            dropout_p=0.0,
            is_causal=True,
        )

        attn_output = attn_output.transpose(1, 2).contiguous()
        attn_output = attn_output.view(bsz, q_len, -1)

        attn_output = transformer_block.self_attn.o_proj(attn_output)

        hidden_states = residual + attn_output

        # Fully Connected
        residual = hidden_states
        hidden_states = transformer_block.post_attention_layernorm(hidden_states)
        hidden_states = transformer_block.mlp(hidden_states)
        hidden_states = residual + hidden_states

    return tuple(blended_kv)
```

Here the argument old_kv refers to the concatenated KV caches of each document. The golden_kv refers to the KV cache that you encode the context as a whole in one forward pass. The correct position ids, same as in the golden_kv, is used when obtaining each part of the old_kv, thus making sure the position information is aligned and avoid the rope scaling problem.

**Comment by lhr-30** [comment]
Thank you for sharing. Would it be possible for you to release the test code for CacheBlend? In the KVLink paper, the reported accuracy results for CacheBlend don’t seem to match the original claims, and I’m currently trying to better understand the accuracy aspect. Having access to the test code would be extremely helpful. Thanks again!

**Comment by nechamab1** [comment]
Hi @KimperYang, thanks for sharing this!
Could you please provide more details on how you generated the `old_kv`?
